test/e2e/upgrade/alert/alert.go

```diff
@@ -52,50 +43,56 @@ func (t *UpgradeTest) Setup(f *framework.Framework) {
 	framework.Logf("Post-upgrade alert test setup complete")
 }
 
-// Test checks if any critical alerts are firing.
+// Test checks if alerts are firing at various points during upgrade.
 func (t *UpgradeTest) Test(f *framework.Framework, done <-chan struct{}, upgrade upgrades.UpgradeType) {
-	g.By("Checking for critical alerts")
+	g.By("Checking for alerts")
 
-	// Recover current test if it fails so test suite can complete
-	defer g.GinkgoRecover()
+	start := time.Now()
 
 	// Block until upgrade is done
-	g.By("Waiting for upgrade to finish before checking for critical alerts")
+	g.By("Waiting for upgrade to finish before checking for alerts")
 	<-done
 
-	ctx, cancel := context.WithCancel(context.Background())
+	// Additonal delay after upgrade completion to allow pending alerts to settle
+	g.By("Waiting before checking for alerts")
+	time.Sleep(1 * time.Minute)
 
-	// Additonal delay after upgrade completion
-	g.By("Waiting before checking for critical alerts")
-	time.Sleep(alertCheckSleep)
-	cancel()
-
-	if exutil.TolerateVersionSkewInTests() {
-		e2eskipper.Skipf("Test is disabled to allow cluster components to have different versions, and skewed versions trigger multiple other alerts")
-	}
-	t.oc.SetupProject()
-	ns := t.oc.Namespace()
+	ns := t.oc.SetupNamespace()
 	execPod := exutil.CreateExecPodOrFail(t.oc.AdminKubeClient(), ns, "execpod")
 	defer func() {
-		t.oc.AdminKubeClient().CoreV1().Pods(ns).Delete(ctx, execPod.Name, *metav1.NewDeleteOptions(1))
+		t.oc.AdminKubeClient().CoreV1().Pods(ns).Delete(context.Background(), execPod.Name, *metav1.NewDeleteOptions(1))
 	}()
 
-	// Query to check if Prometheus has been up and running for entire post-upgrade
-	// period by verifying Watchdog alert has been in firing state
-	watchdogQuery := fmt.Sprintf(`count_over_time(ALERTS{alertstate="firing",alertname="Watchdog", severity="none"}[%dm])`, alertCheckSleepMinutes)
+	testDuration := time.Now().Sub(start).Truncate(time.Second)
+
+	// The watchdog alert should be firing continuously during the whole upgrade except when prometheus
+	// is stopped, which the joined query represents by looking at a second metric that is always reported
+	// when the cluster is running. We do not use the prometheus_* metrics because they can't be reviewed
+	// historically.
+	// TODO: switch to using thanos querier and verify that we can query the entire history of the upgrade
+	watchdogQuery := fmt.Sprintf(`sum_over_time((ALERTS{alertstate="firing",alertname="Watchdog",severity="none"} * scalar(absent(max(openshift:prometheus_tsdb_head_samples_appended_total:sum*0+1))))[%s:1s]) > 0`, testDuration)
```

Comment:
Poking at [the CI job][1] with PromeCIeus: update [from 2021-02-22T20:35:58Z to 2021-02-22T21:40:29Z][2], and Prom was down from 21:24 to 21:26:

![image](https://user-images.githubusercontent.com/209920/108777537-cc2b8080-7518-11eb-9587-f765e1f30fb4.png)

Is Prom going down like that mid-update acceptable?  If it is, I think we should drop the `Watchdog` guard.  If not, I think we should drop the `scalar(absent(...` business.

[1]: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/25904/pull-ci-openshift-origin-master-e2e-gcp-upgrade/1363938413538971648
[2]: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/25904/pull-ci-openshift-origin-master-e2e-gcp-upgrade/1363938413538971648/artifacts/e2e-gcp-upgrade/gather-extra/artifacts/clusterversion.json

Reply:
This is the thanos-querier change i mentioned in slack.